docs/controller-guidelines.md

```diff
@@ -539,189 +543,178 @@ A controller should define and export a type union that holds all of its events.
+✅ **`FooControllerEvents` is passed as the 2nd type parameter (assuming no external events)**
 
 ```typescript
 export type FooControllerEvents =
   | FooControllerMessageReceivedEvent
   | FooControllerNotificationAddedEvent;
 
-export type FooControllerMessenger = RestrictedMessenger<
+export type FooControllerMessenger = Messenger<
   'FooController',
-  never,
-  FooControllerEvents,
-  never,
-  FooControllerEvents['type']
->;
-```
-
-✅ **`never` is passed as the 5th type parameter (assuming no external events)**
-
-```typescript
-export type FooControllerEvents =
-  | FooControllerMessageReceivedEvent
-  | FooControllerNotificationAddedEvent;
-
-export type FooControllerMessenger = RestrictedMessenger<
-  'FooController',
-  never,
-  FooControllerEvents,
-  never,
-  never
+  FooControllerActions,
+  FooControllerEvents
 >;
 ```
 
 ## Define, but do not export, a type union for external action types
 
-A controller may wish to call actions defined by other controllers, and therefore will need to define them in the messenger's allowlist.
+A controller may wish to call actions defined by other controllers, and therefore will need to include them in the controller messenger's type definition.
 
-In this case, the controller should group these types into a type union so that they can be easily passed to the `RestrictedMessenger` type. However, it should not export this type, as it would then be re-exporting types that another package has already exported.
+In this case, the controller should group these types into a type union so that they can be easily passed to the `Messenger` type. However, it should not export this type, as it would then be re-exporting types that another package has already exported.
 
 The name of this type should be `AllowedActions`.
 
-This type should not only be passed to `RestrictedMessenger` as the 2nd type parameter, but should also be included in its 4th type parameter.
+This type should be passed to `Messenger` as part of the 2nd type parameter, in a type union with internal actions.
 
-🚫 **`never` is passed as the 4th type parameter**
+🚫 **`AllowedActions` is exported**
 
 ```typescript
+/* === packages/foo-controller/src/FooController.ts === */
+
+export type FooControllerActions =
+  | FooControllerUpdateCurrencyAction
+  | FooControllerUpdateRatesAction;
+
 export type AllowedActions =
   | BarControllerDoSomethingAction
   | BarControllerDoSomethingElseAction;
 
-export type FooControllerMessenger = RestrictedMessenger<
+export type FooControllerMessenger = Messenger<
   'FooController',
-  AllowedActions,
-  never,
-  never,
-  never
+  FooControllerActions | AllowedActions,
+  FooControllerEvents
 >;
+
+/* === packages/foo-controller/src/index.ts === */
+
+export type { AllowedActions } from '@metamask/foo-controller';
 ```
 
-🚫 **`AllowedActions['type']` is passed as the 4th type parameter, but `AllowedActions` is exported**
+🚫 **External actions are included in controller action type**
```

Comment:
Perhaps this anti-pattern should be moved up to "Define and export a type union for internal action types"?